pkg/kubelet/custommetrics/impl.go

```diff
@@ -0,0 +1,156 @@
+
+	CustomMetricsDefinitionContainerPath = "/etc/" + CustomMetricsDefinitionContainerFile
+)
+
+// CustomMetricsEndpoint is a structure defining endpoint for pod's custom metric.
+// It is parsed from JSON from the value of the annotation with CustomMetricsAnnotationKey.
+// More info: http://releases.k8s.io/HEAD/docs/proposals/custom-metrics.md#api
+type CustomMetricsEndpoint struct {
+	API   string   `json:"api",omitempty`
+	Path  string   `json:"path",omitempty`
+	Port  int      `json:"port"`
+	Names []string `json:"names"`
+}
+
+func MakeCustomMetricsMountIfRequired(
+	pod *api.Pod, container *api.Container, podContainerDir string, maxCustomMetricsPerPod int) (*kubecontainer.Mount, error) {
+	if definition, ok := pod.Annotations[CustomMetricsAnnotationKey]; ok {
+		customMetricsMount, err := makeCustomMetricsMount(
+			container, podContainerDir, pod.Status.PodIP, definition, maxCustomMetricsPerPod)
+		if err != nil {
+			return nil, err
+		}
+		return customMetricsMount, nil
+	}
+	return nil, nil
+}
+
+func VerifyCustomMetricsAnnotation(pod *api.Pod, maxCustomMetricsPerPod int) error {
+	if definition, ok := pod.Annotations[CustomMetricsAnnotationKey]; ok {
+		for _, container := range pod.Spec.Containers {
+			if _, err := getCustomMetricsEndpointsForContainer(&container, definition, maxCustomMetricsPerPod); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func makeCustomMetricsMount(container *api.Container, podContainerDir, podIP, definition string, maxCustomMetricsPerPod int) (*kubecontainer.Mount, error) {
+	endpoints, err := getCustomMetricsEndpointsForContainer(container, definition, maxCustomMetricsPerPod)
+	if err != nil {
+		return nil, err
+	}
+	if len(endpoints) == 0 {
+		return nil, nil
+	}
+
+	if err := os.MkdirAll(podContainerDir, 0750); err != nil {
+		return nil, fmt.Errorf("error creating pod-container directory: %v", err)
+	}
+	hostsFilePath := path.Join(podContainerDir, CustomMetricsDefinitionContainerFile)
+	if err := ensureCustomMetricsFile(hostsFilePath, podIP, endpoints); err != nil {
+		return nil, err
+	}
+	return &kubecontainer.Mount{
+		Name:          "k8s-custom-metrics",
+		ContainerPath: CustomMetricsDefinitionContainerPath,
+		HostPath:      hostsFilePath,
+		ReadOnly:      false,
+	}, nil
+}
+
+func getCustomMetricsEndpointsForContainer(container *api.Container, definition string, maxCustomMetricsPerPod int) ([]CustomMetricsEndpoint, error) {
+	var endpoints []CustomMetricsEndpoint
+	err := json.Unmarshal([]byte(definition), &endpoints)
+	if err != nil {
+		return nil, fmt.Errorf("error while parsing custom metrics definition %q: %v", definition, err)
+	}
+
+	numCustomMetrics := 0
+	for _, endpoint := range endpoints {
+		if len(endpoint.Names) == 0 {
+			return nil, fmt.Errorf("custom metrics endpoint definition cannot contain empty names: +%v", endpoint)
+		}
+		numCustomMetrics += len(endpoint.Names)
+	}
+	if numCustomMetrics > maxCustomMetricsPerPod {
```

Comment:
Ahh, I misread the code above: VerifyCustomMetricsAnnotation also call getCustomMetricsEndpointsForContainer before makeCustomMetricsMount call the same validation code here.

Then why not move this logic and other validation piece to Verify* method? If validated correctly, we has endpoint for makeCustomMetricsMount?